### ISSUE TITLE
Question: How to serialize a StateMachine object to JSON 

### ISSUE THREAD

**Issue by bhaidar** [dotnet-state-machine/stateless#194]
Hi,
Is there a way to serialize a StateMachine to JSON? I would like to store all my StateMachines in Database in the form of JSON strings.

Thanks
Bilal

**Comment by HenningNT** [comment]
Have a look at #104, it should give you an idea about how to get started.

**Comment by HenningNT** [comment]
I think we maybe should add an example project to do JSON export, or add something to the readme?

**Comment by HenningNT** [comment]
@bhaidar could you please have a look at the pull request, does it clarify, or do you think it could use a bit more work?

**Comment by bhaidar** [comment]
@HenningNT 
Thanks for the update. I've been using it now for like 2 weeks and things going fine with an ASP.NET Core app. Working as expected, etc.

My initial idea was to be able to provide my users a UI Designer, where they can drag n drop states and connections between them. I would then configure the StateMachine with all states and transitions and store in the DB.

I believe the ToJson and FromJson are perfect in this regard. However, still something missing which is ability to configure StateMachine based on an input or so to allow offering a (Designer or tool). I am not sure if this is curently doable or not. Maybe you can illustrate more on it.

This might not be the original plan of the library, but coming from a background where we use 3rd party engines for Workflow (Stored in Db, etc.), I am trying to make use of this library in the same regard.

In summary, the Json feature is great and useful. I can now store the whole StateMachine in Db. Maybe a step forward :)


Thanks
Bilal H.

**Comment by HenningNT** [comment]
Glad I could help!

I did something somewhat similar to what you describe; I created an editor where the user could configure a warm-up sequence for an HV generator. The sequence was built up by stages, which could be combined in any order. Each stage was a state machine and the sequence was strictly linear. The user could only configure parameters for each stage, and the exit condition. I created a graph for the sequence, and corresponding UI controls for each stage for configuring the stage, but I regret doing it that way. I could have generated the control on the fly by inspecting either a config file or querying a static class for what sort of parameters the class needed configuring.
The user didn't select or know about the trigger names or their handlers, so the sequence generation was fairly straightforward.
Since Stateless can use anything for its states and events it is perfectly possible to configure a state machine from a text file. I guess the hardest "problem" is hooking up the actions...

**Comment by bhaidar** [comment]
@HenningNT 
Thanks for sharing the info. Wouldn't be too much to compile, etc.?

I thought of using one Handler for all triggers. Then, use some logic to direct triggers accordingly since I believe I can use a String value as a Trigger Name. I am going to dig more into this and see if that is doable with the current Stateless design and API.

Thanks
Bilal

**Comment by psibernetic** [comment]
I had a similar need, so I have been working on a new state machine project [REstate](https://github.com/psibr/REstate) that keeps portability as a first-class concern. The API surface area is significantly narrower than Stateless to reduce complexity, so it may or may not be a fit for your designs. Make sure to check out the wiki, examples, and the [ConsoleConnector](https://github.com/psibr/REstate/tree/master/src/REstate/Engine/Connectors/Console) as an example of the EntryActions and Guards.
